accepted/2021/json-polymorphism.md

```diff
@@ -0,0 +1,275 @@
+
+* Polymorphic serialization applied indiscriminately can result in unintended data leaks, 
+  since properties of unexpected derived types may end up written on the wire.
+* Polymorphic deserialization can be vulnerable when deserializing untrusted data,
+  in certain cases leading to remote code execution attacks.
+
+## Introduction
+
+The proposed design for polymorphic serialization in System.Text.Json can be split into two
+largely orthogonal features:
+
+1.  Simple Polymorphic serialization: extends the existing serialization infrastructure for `object` types to 
+    arbitrary classes that can be specified by the user. It trivially dispatches to the converter corresponding 
+    to the runtime type without emitting any metadata on the wire and does not provide any provision for 
+    polymorphic deserialization.
+2.  Polymorphism with type discriminators ("tagged polymorphism"): classes can be serialized and deserialized
+    polymorphically by emitting a type discriminator ("tag") on the wire. Users must explicitly associate each
+    supported subtype of a given declared type with a string identifier.
+
+## Simple Polymorphic Serialization
+
+Consider the following type hierarchy:
+```csharp
+public class Foo
+{
+    public int A { get; set; }
+}
+
+public class Bar : Foo
+{
+    public int B { get; set; }
+}
+
+public class Baz : Bar
+{
+    public int C { get; set; }
+}
+```
+Currently, when serializing a `Bar` instance as type `Foo`
+the serializer will apply the JSON schema derived from the type `Foo`:
+```csharp
+Foo foo1 = new Foo { A = 1 };
+Foo foo2 = new Bar { A = 1, B = 2 };
+Foo foo3 = new Baz { A = 1, B = 2, C = 3 };
+
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo3); // { "A" : 1 }
+```
+Under the new proposal we can change this behaviour by annotating 
+the base class (or interface) with the `JsonPolymorphicType` attribute:
+```csharp
+[JsonPolymorphicType]
+public class Foo
+{
+  ...
+}
+```
+which will result in the above values now being serialized as follows:
+```csharp
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo3); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+Note that the `JsonPolymorphicType` attribute is not inherited by derived types.
+In the above example `Bar` inherits from `Foo` yet is not polymorphic in its own right:
+```csharp
+Bar bar = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Bar>(bar); // { "A" : 1, "B" : 2 }
+```
+If annotating the base class with an attribute is not possible,
+polymorphism can alternatively be opted in for a type using the
+new `JsonSerializerOptions.SupportedPolymorphicTypes` predicate:
+```csharp
+public class JsonSerializerOptions
+{
+  public Func<Type, bool> SupportedPolymorphicTypes { get; set; }
```

Comment:
I assume the result is cached on the corresponding JsonTypeInfo, in order to prevent potential perf issues with a complex func.

Reply:
Correct, cf. https://github.com/dotnet/runtime/pull/53882/files#diff-8d841bf86f2a9da83758276bd17f55d002ee5b55f58b72da658df351a7c22bc0R553